### PR TITLE
[Group 4] Enable nullable annotations for `Microsoft.Extensions.Logging.EventSource`

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to https://github.com/dotnet/runtime/issues/43605

Questions:
- `Arguments` can be null here
https://github.com/dotnet/runtime/blob/731d936d37ce760327f96337e1ff3dbcb8078680/src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs#L303
- I'm unable to build this repository for the past couple of weeks, installing .Net 7 Preview 3 from `main` doesn't help. Is there something I'm missing?
```
CS0433	The type 'Unsafe' exists in both 'System.Runtime.CompilerServices.Unsafe, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'	Microsoft.Extensions.Logging.EventSource (net7.0)
```
image

### CODE REVIEW

**Comment by eerhardt** [comment]
> I'm unable to build this repository for the pas couple of weeks, installing .Net 7 Preview 3 from main doesn't help. Is there something I'm missing?

I would try cleaning my repo: `git clean -xdf` and then rebuild from the root `.\build.cmd libs+clr -rc Release`. If you see have errors, let me know.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/ref/Microsoft.Extensions.Logging.EventSource.cs [hunk 1]
```diff
@@ -17,7 +17,7 @@ namespace Microsoft.Extensions.Logging.EventSource
     public partial class EventSourceLoggerProvider : Microsoft.Extensions.Logging.ILoggerProvider, System.IDisposable
     {
         public EventSourceLoggerProvider(Microsoft.Extensions.Logging.EventSource.LoggingEventSource eventSource) { }
-        public Microsoft.Extensions.Logging.ILogger CreateLogger(string categoryName) { throw null; }
+        public Microsoft.Extensions.Logging.ILogger CreateLogger(string? categoryName) { throw null; }
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/ref/Microsoft.Extensions.Logging.EventSource.cs:20]: `ILoggerProvider` has categoryName as non-nullable. So this should be `string categoryName`.

We don't want to allow this to be null.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs [hunk 2]
```diff
@@ -451,7 +451,7 @@ private static bool TryParseLevel(LogLevel defaultLevel, string levelString, out
                     {
                         return false;
                     }
-                    if (!(LogLevel.Trace <= (LogLevel)level && (LogLevel)level <= LogLevel.None))
+                    if ((LogLevel)level is not (>= LogLevel.Trace and <= LogLevel.None))
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:454]: I don't think this is more readable than the current code. Can you revert this change?

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLoggerProvider.cs [hunk 3]
```diff
@@ -27,7 +28,8 @@ public EventSourceLoggerProvider(LoggingEventSource eventSource!!)
         }
 
         /// <inheritdoc />
-        public ILogger CreateLogger(string categoryName)
+        [MemberNotNull(nameof(_loggers))]
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/EventSourceLoggerProvider.cs:31]: Does this attribute help any compiler warnings? If not, I would remove it.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs]: Yeah, both are not really readable, but VS suggested it, and I thought there might be some performance improvement. Reverted.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/ref/Microsoft.Extensions.Logging.EventSource.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/ref/Microsoft.Extensions.Logging.EventSource.cs]: Should `EventSourceLogger.CategoryName` be nullable then? Maybe it doesn't really matter sines it's an internal API, and nothing breaks when it's nullable, but still.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Logging.EventSource/ref/Microsoft.Extensions.Logging.EventSource.cs]: > Should EventSourceLogger.CategoryName be nullable then? 

I would say `no`. If nothing ever creates it with a `null` name, then I wouldn't make it nullable.

File: src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs [hunk 4]
```diff
@@ -298,9 +298,9 @@ internal unsafe void ActivityJsonStop(int ID, int FactoryID, string LoggerName)
         /// <inheritdoc />
         protected override void OnEventCommand(EventCommandEventArgs command)
         {
-            if (command.Command == EventCommand.Update || command.Command == EventCommand.Enable)
+            if (command.Command is EventCommand.Update or EventCommand.Enable)
             {
-                if (!command.Arguments.TryGetValue("FilterSpecs", out string filterSpec))
+                if (!command.Arguments!.TryGetValue("FilterSpecs", out string? filterSpec))
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.EventSource/src/LoggingEventSource.cs:303]: This is the only place that concerns me. But if it hasn't been a problem, I guess we can just suppress the null check here. Looking around I see similar code elsewhere:

https://github.com/dotnet/runtime/blob/c5d40c9e703fd257db1b26ef4fd1399bbae73ab0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L393-L397

https://github.com/dotnet/runtime/blob/c5d40c9e703fd257db1b26ef4fd1399bbae73ab0/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/CounterGroup.cs#L53-L59

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution here, @maxkoshevoi!

**Comment by danmoseley** [comment]
+1 we appreciate your hard work on all these annotations @maxkoshevoi .
